### PR TITLE
[Dygraph] Add MoE model pretrain and support MoE+DP

### PR DESCRIPTION
**Pull request by haohongxiang**
### PR types

Others
### PR changes

Others
### Description

[Dygraph] Add MoE model and support MoE+DP

### CODE REVIEW

File: examples/language_model/moe/dygraph/result.txt [hunk 1]
```diff
@@ -0,0 +1,4004 @@
+grep: warning: GREP_OPTIONS is deprecated; please use an alias or script
```
Comment (ZHUI) [inline examples/language_model/moe/dygraph/result.txt]: delete?

File: examples/language_model/moe/dygraph/framework/adamw.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+#
```
Comment (ZHUI) [inline examples/language_model/moe/dygraph/framework/adamw.py:2]: why need this ?

File: examples/language_model/moe/dygraph/framework/storage_process.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (ZHUI) [inline examples/language_model/moe/dygraph/framework/storage_process.py:1]: 这部分的作用是，以后会到框架中吗？

File: examples/language_model/moe/dygraph/utils.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
+import os
```
Comment (ZHUI) [inline examples/language_model/moe/dygraph/utils.py:15]: copyright

File: examples/language_model/moe/dygraph/args.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (ZHUI) [inline examples/language_model/moe/dygraph/args.py:1]: Please add some readme

File: examples/language_model/moe/dygraph/result.txt (hunk 1 above)
Comment (haohongxiang) [reply examples/language_model/moe/dygraph/result.txt]: Done

File: examples/language_model/moe/dygraph/utils.py (hunk 4 above)
Comment (haohongxiang) [reply examples/language_model/moe/dygraph/utils.py:15]: Done

File: examples/language_model/moe/dygraph/framework/storage_process.py (hunk 3 above)
Comment (haohongxiang) [reply examples/language_model/moe/dygraph/framework/storage_process.py:1]: 这部分代码是基于框架的二次开发，不合入框架repo

File: examples/language_model/moe/dygraph/checkpointing.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
+import os
```
Comment (ZHUI) [inline examples/language_model/moe/dygraph/checkpointing.py:1]: copyright

**Review by ZHUI — APPROVED** [review]
下个PR修复吧，还有一些Copyright缺失之类的

Copyright (c) 2021 -> Copyright (c) 2022

**Pull request merged**